File: src/components/modules/banners.js
const http = window.__TAURI__.http;
const invoke = window.__TAURI__.invoke;
const path = window.__TAURI__.path;
const tauri = window.__TAURI__.tauri;
const { sha256 } = require("../modules/sha256");

async function getBanners(games) {
  games = games.filter((x) => !["CustomGame"].includes(x.LauncherName));
  const bannerBasePath = (await path.appDir()) + "cache/games/banners";
  const readBanners = await invoke('read_dir', { dirPath: bannerBasePath })

  let alreadyProcessed = false;
  let existingProcessed = 0;

  for (let i = 0; i < games.length; i++) {
    if (
      readBanners.includes(
        `${sha256(games[i].DisplayName.replaceAll(" ", "_"))}.png`
      )
    ) {
      existingProcessed++;
      if (existingProcessed === games.length) {
        alreadyProcessed = true;
      }
    }
  }
  if (alreadyProcessed === true) {
    return console.log(
      "%c[BANNER] " + "Banners are already loaded. Skipping.",
      "color:blue"
    );
  }

  const arr = [];
  for (let i = 0; i < games.length; i++) {
    arr.push(
      (async () => {
        switch (games[i].LauncherName) {
          case "Steam": {
            return `https://cdn.akamai.steamstatic.com/steam/apps/${games[i].GameID}/library_600x900.jpg`;
          }
          case "RiotGames": {
            return "https://valorant-config.fr/wp-content/uploads/2020/05/7d604cf06abf5866f5f3a2fbd0deacf9-200x300.png";
          }
          case "Minecraft": {
            return "https://i.imgur.com/PJFx3U2.jpg";
          }
          case "FiveM": {
            return "https://logos-world.net/wp-content/uploads/2021/03/FiveM-Symbol.png";
          }
          case "Lunar": {
            return "https://www.lunarclient.com/assets/img/default-twitter-icon.webp";
          }
          case "Lutris": {
            if (
              games[i].DisplayName.replaceAll(" ", "_") === "Epic_Games_Store"
            )
              return "https://pcper.com/wp-content/uploads/2021/02/epic-games-store.png";
            if (
              games[i].DisplayName.replaceAll(" ", "_") ===
              "Rockstar_Games_Launcher"
            )
              return "https://cdn.player.one/sites/player.one/files/2019/08/26/rockstar-games.png";
            return;
          }
          case "XboxGames": {
            return games[i].Banner;
          }
          case "Osu": {
            return "https://cdn2.steamgriddb.com/file/sgdb-cdn/grid/a5d7420f9fdc41087377b4d58c5fe94b.png";
          }
          case "EpicGames": {
            const fetchEpicGame = await http.fetch(
              `https://api.rawg.io/api/games?key=f8854c401fed44b89f4e1e4faa56ccc8&search=${games[
                i
              ].DisplayName.replaceAll(
                " ",
                "-"
              )}&search_exact&search_precise&stores=11`,
              {
                method: "GET",
                mode: "no-cors",
                headers: {
                  Accept: "application/json",
                  "Content-Type": "application/json",
                },
              }
            );
            if (!fetchEpicGame.data.results[0]) break;
            return (
              fetchEpicGame.data.results[0].background_image.slice(0, 27) +
              "/crop/600/400" +
              fetchEpicGame.data.results[0].background_image.slice(27)
            );
          }
          case "Uplay": {
            const fetchUplay = await http.fetch(
              `https://api.rawg.io/api/games?key=f8854c401fed44b89f4e1e4faa56ccc8&search=${games[
                i
              ].DisplayName.replaceAll(" ", "-")}&search_exact&search_precise`,
              {
                method: "GET",
                mode: "no-cors",
                headers: {
                  Accept: "application/json",
                  "Content-Type": "application/json",
                },
              }
            );
            if (!fetchUplay.data.results[0]) break;
            return (
              fetchUplay.data.results[0].background_image.slice(0, 27) +
              "/crop/600/400" +
              fetchUplay.data.results[0].background_image.slice(27)
            );
          }
          case "Rockstar": {
            return `https://media-rockstargames-com.akamaized.net/rockstargames-newsite/img/global/games/fob/640/${games[i].bannerId}.jpg`;
          }
        }
      })()
    );
  }

  cacheBanners(
    games,
    arr.filter((x) => x)
  );
  return arr;
}

async function cacheBanners(data, res) {
  const bannerBasePath = (await path.appDir()) + "cache/games/banners";

  if (data?.length === 0) {
    return console.log("%c[BANNER] " + "No games to process", "color:blue");
  }

  let fetchProcessed = 0;

  if (res.length === 0) {
    return console.log("%c[BANNER] " + "No banners to load.", "color:blue");
  }

  res
    .filter(async (x) => (await x)?.startsWith("http"))
    .forEach(async (x, i) => {
      await http
        .fetch(await x, {
          method: "GET",
          headers: {
            "Content-Type": "text/plain; charset=utf-8",
          },
          responseType: 3,
        })
        .then(async (response) => {
          if (response.status === 404 && data[i].LauncherName === "Lutris")
            return;
          await invoke("write_binary_file", {
            filePath:
              bannerBasePath +
              `/${sha256(data[i].DisplayName.replaceAll(" ", "_"))}.png`,
            fileContent: response.data,
          });
          document
            .getElementById(
              `game-div-${data[i].DisplayName.replaceAll(" ", "_")}`
            )
            ?.firstElementChild?.setAttribute(
              "src",
              tauri.convertFileSrc(
                bannerBasePath +
                  `/${sha256(data[i].DisplayName.replaceAll(" ", "_"))}.png`
              )
            );
        })
        .catch((e) => console.error(e));

      fetchProcessed++;
      if (fetchProcessed === data.length) {
        return console.log(
          "%c[BANNER] " + "Just finished processing banners.",
          "color:blue"
        );
      }
    });
}

export { getBanners };

File: src-tauri/src/modules/banners.rs
use std::{
    fs::File,
    io::{copy, Cursor},
};

use reqwest::header::{ACCEPT, CONTENT_TYPE};
use serde::Deserialize;

use crate::{
    operations::{custom_fs::d_f_exists, misc::sha256},
    CONFIG_DIR,
};

pub async fn get_banner(
    display_name: &str,
    game_id: &str,
    launcher_name: &str,
    url: &str,
) -> String {
    match launcher_name {
        "Steam" => {
            return fetch_banner(
                format!(
                    "https://cdn.akamai.steamstatic.com/steam/apps/{}/library_600x900.jpg",
                    game_id
                ),
                display_name,
            )
            .await;
        }
        "RiotGames" => {
            return fetch_banner(format!("https://valorant-config.fr/wp-content/uploads/2020/05/7d604cf06abf5866f5f3a2fbd0deacf9-200x300.png"), display_name).await;
        }
        "Minecraft" => {
            return fetch_banner(format!("https://i.imgur.com/PJFx3U2.jpg"), display_name).await;
        }
        "FiveM" => {
            return fetch_banner(
                format!("https://logos-world.net/wp-content/uploads/2021/03/FiveM-Symbol.png"),
                display_name,
            )
            .await;
        }
        "Lunar" => {
            return fetch_banner(
                format!(
                    "https://pbs.twimg.com/profile_images/1608698913476812801/uLTLhANK_400x400.jpg"
                ),
                display_name,
            )
            .await;
        }
        "Lutris" => {
            if display_name.replace(" ", "_") == "Epic_Games_Store" {
                return fetch_banner(
                    format!("https://pcper.com/wp-content/uploads/2021/02/epic-games-store.png"),
                    display_name,
                )
                .await;
            }
            if display_name.replace(" ", "_") == "Rockstar_Games_Launcher" {
                return fetch_banner(format!("https://cdn.player.one/sites/player.one/files/2019/08/26/rockstar-games.png"), display_name).await;
            }
            return String::new();
        }
        // "XboxGames" => {
        //     return fetch_banner(url.to_string(), display_name).await;
        // },
        "Osu" => {
            return fetch_banner(format!("https://cdn2.steamgriddb.com/file/sgdb-cdn/grid/a5d7420f9fdc41087377b4d58c5fe94b.png"), display_name).await;
        }
        "EpicGames" => {
            return rawg_fetch_banner(display_name).await;
        }
        "Uplay" => {
            return rawg_fetch_banner(display_name).await;
        }
        "RockstarGames" => {
            return fetch_banner(format!("https://media-rockstargames-com.akamaized.net/rockstargames-newsite/img/global/games/fob/640/{}.jpg", game_id), display_name).await;
        }
        "GOG" => {
            return fetch_banner(url.to_string(), display_name).await;
        }

        _ => return String::new(),
    }
}

async fn fetch_banner(url: String, display_name: &str) -> String {
    let banners_dir = format!("{}/cache/games/banners", CONFIG_DIR.lock().unwrap());

    let dispsha256 = sha256(display_name.replace(" ", "_")).await.unwrap();
    let file_path: String = format!("{}/{}.png", banners_dir, dispsha256);
    if d_f_exists(&file_path).await.unwrap() {
        return file_path;
    }

    let response = reqwest::get(url).await.unwrap();
    let mut file = File::create(&file_path).unwrap();

    let mut content = Cursor::new(response.bytes().await.unwrap());
    copy(&mut content, &mut file).unwrap();

    return file_path;
}

#[derive(Deserialize, Debug)]
struct RAWGApi {
    results: Vec<GameWithinRawg>,
}

#[derive(Deserialize, Debug)]
struct GameWithinRawg {
    background_image: String,
}

async fn rawg_fetch_banner(display_name: &str) -> String {
    let banners_dir = format!("{}/cache/games/banners", CONFIG_DIR.lock().unwrap());

    let dispsha256 = sha256(display_name.replace(" ", "_")).await.unwrap();
    let file_path: String = format!("{}/{}.png", banners_dir, dispsha256);
    if d_f_exists(&file_path).await.unwrap() {
        return file_path;
    }

    let client = reqwest::Client::new();
    let response = client
        .get(format!("https://api.rawg.io/api/games?key=f8854c401fed44b89f4e1e4faa56ccc8&search={}&search_exact&search_precise&stores=11", display_name.replace(" ", "-").replace("_", "-")))
        .header(CONTENT_TYPE, "application/json")
        .header(ACCEPT, "application/json")
        .send()
        .await
        .unwrap();

    match response.status() {
        reqwest::StatusCode::OK => {
            let background_image_url =
                match response.json::<RAWGApi>().await.unwrap().results.get(0) {
                    Some(result) => result.background_image.clone(),
                    None => return String::new(),
                };

            let banner_response = match reqwest::get(&background_image_url).await {
                Ok(response) => response,
                Err(err) => {
                    eprintln!("Error fetching background image: {}", err);
                    return String::new();
                }
            };
            let mut file = File::create(&file_path).unwrap();

            let mut content = Cursor::new(banner_response.bytes().await.unwrap());
            copy(&mut content, &mut file).unwrap();

            return file_path;
        }
        _ => {
            println!("Something unexpected happened... Trace -> init_reqwest");
            return String::new();
        }
    }
}
